aten/src/ATen/native/TensorShape.cpp

```diff
@@ -123,11 +123,11 @@ TORCH_PRECOMPUTE_META_FUNC(cat)(ITensorListRef tensors, int64_t dim) {
     size_t size_at_dim = 0;
     for (const auto i : c10::irange(materialized.size())) {
       const Tensor& t = materialized[i];
+      all_same_dtype = all_same_dtype && out_dtype == t.scalar_type();
```

Comment:
We should not skip checking dtype even if the tensor has 0 dim size. (As specialised kernel down the path may not work if all tensors are not same dtype eg. CUDA).

Reply:
Wow. You are right! CUDA implementation does iterate (as far as I can understand, unnecessarily) over all tensors. Nice catch!